Add Django OAuth login with Pywikibot integration

sergyDwhiz opened the pull request:
This pull request integrates Django Social Auth with the MediaWiki OAuth 1.0a backend, enabling users to authenticate using their Wikimedia accounts.

Changes made include: 
- Integrated and configured social-auth-app-django with the MediaWiki backend.
- Added a “Login with Wikimedia” button to the top-right navigation bar (visible only when OAuth is enabled).
- Implemented a robust logout flow that fully clears user sessions.
- Introduced a helper function which retrieves OAuth tokens from Django Social Auth and supplies them to Pywikibot for authenticated bot operations.
- Added an OAUTH_ENABLED environment variable for backward compatibility — when disabled, the application retains its previous authentication behavior.
- Updated all relevant settings, URLs, views, and templates to support the new authentication flow.
- Included a comprehensive testing guide to facilitate validation of the integration.

Addresses: https://github.com/Wikimedia-Suomi/PendingChangesBot-ng/issues/117

sergyDwhiz committed:
feat: Implement Django OAuth login with Pywikibot integration

- Add Django Social Auth with MediaWiki OAuth 1.0a backend
- Configure OAuth callback URL with trailing slash
- Add login/logout views and URL routes
- Update UI with login button in top right corner
- Implement configure_pywikibot_oauth() helper function

sergyDwhiz committed:
docs: Update OAuth configuration for production and add testing status

sergyDwhiz committed:
Resolved conflicts in app/reviews/views.py

zache-fi commented:
Hi, try to register it at [meta.wikimedia.beta.wmflabs.org/](https://meta.wikimedia.beta.wmflabs.org/wiki/Special:OAuthConsumerRegistration/propose)  and use it as endpoint. It doesn't work for superset, but I can  approve ouths there so registering and finding correct value can be done trial and error.

zache-fi commented:
Note https://meta.wikimedia.beta.wmcloud.org doesn't use wikimedias unified login so you need to register separate account there. 

**Oauth consumer registration**
So I used these values when i registered the Oauth consumer in https://meta.wikimedia.beta.wmcloud.org/wiki/Special:OAuthConsumerRegistration/propose 

- Application name: `PendingChangesBot test 2`
- Consumer version: `1.0`
- Application description: `PendingChangesBot is FlaggedRevs  automatic review bot ( https://github.com/Wikimedia- Suomi/PendingChangesBot-ng )`
- This consumer is for use only by Zache-test: NOT SELECTED
- OAuth "callback" URL: `http://127.0.0.1:8000/`
- Allow consumer to specify a callback in requests and use "callback" URL above as a required prefix.: SELECTED
- Contact email address: MY EMAIL ADDRESS
- Applicable project: `*`
- Types of grants being requested: `Request authorization for specific permissions.`
- Applicable grants:
   - `Edit existing pages`
   - `Create, edit, and move pages`
   - `Patrol changes to pages`
   - `Rollback changes to pages`
   
**settings.py**   
```
# MediaWiki OAuth 1.0a settings
# For development: use meta.wikimedia.org (beta is often blocked)
# For production: use meta.wikimedia.org
SOCIAL_AUTH_MEDIAWIKI_URL = "https://meta.wikimedia.beta.wmcloud.org/w/index.php"
SOCIAL_AUTH_MEDIAWIKI_KEY = "CONSUMER_TOKEN_FROM_OAUTH_REGISTRATION"
SOCIAL_AUTH_MEDIAWIKI_SECRET = "SECRET_TOKEN_FROM_OAUTH_REGISTRATION"
SOCIAL_AUTH_MEDIAWIKI_CALLBACK = "http://127.0.0.1:8000/oauth/complete/mediawiki/"
```

zache-fi commented:
Btw is beta blocked for you?

sergyDwhiz committed:
Merge upstream/main and resolve requirements.txt conflict

- Keep social-auth-app-django for OAuth
- Keep upstream type checking and security tools

sergyDwhiz committed:
docs: Update OAuth to use beta.wmcloud.org per maintainer guidance

- Changed default SOCIAL_AUTH_MEDIAWIKI_URL to meta.wikimedia.beta.wmcloud.org
- Updated OAUTH_TESTING_GUIDE.md with Zache's specific registration instructions

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

sergyDwhiz committed:
feat: Improve OAuth UI and Pywikibot integration per maintainer feedback

- Add top navigation bar with login button in top-right corner
- Save original Wikimedia username in first_name field (avoid normalization)
- Update Pywikibot OAuth config to use beta/production appropriately
- Configure pywikibot.config.usernames based on OAuth endpoint
- Display first_name (original username) in navbar if available
- Fix ruff formatting issues

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

zache-fi commented:
check why the test is failing and fix it

sergyDwhiz committed:
fix: Resolve Django/Vue.js template syntax conflicts

- Replace Vue.js syntax with v-text directive to avoid Django parser conflicts
- Remove verbatim tags as they're no longer needed with v-text

xenacode-art commented:
Hi @sergyDwhiz! 👋

  Excellent work on the OAuth integration! I've reviewed the code and tested it locally. This is a crucial      
  piece of infrastructure for the internship roadmap. Here's my detailed feedback:

  ## ✅ What Works Well

  **Architecture:**
  - Clean separation of OAuth logic using Django Social Auth
  - Smart handling of beta vs production endpoints
  - Backwards compatibility with `OAUTH_ENABLED` flag is great for gradual rollout

  **Pywikibot Integration:**
  - The `get_pywikibot_site_with_oauth()` helper function is elegant
  - Automatically configuring Pywikibot from Django session tokens avoids credential duplication
  - Good error handling when OAuth tokens aren't available

  **UI/UX:**
  - Login button placement in navigation makes sense
  - Preserving original Wikimedia usernames in `first_name` is clever workaround for Django's normalization     

  ## Testing & Observations

  I tested this locally with the following scenarios:

  **Scenario 1: OAuth Login Flow**
  - [x] Login button appears when not authenticated
  - [x] Redirects to Wikimedia OAuth authorization
  - [x] Successfully returns and creates Django session
  - [x] Username displays correctly in navbar

  **Scenario 2: OAuth Disabled Mode**
  - [x] Application works with `OAUTH_ENABLED=False`
  - [x] Previous authentication methods still function
  - [x] No errors in logs

  **Scenario 3: Pywikibot Integration**
  - [x] OAuth tokens passed correctly to Pywikibot
  - [x] Bot operations work with OAuth credentials
  - [ ] ⚠️ Need to verify with actual flaggedrevs operations (requires wiki with FlaggedRevs enabled)

  ##  Minor Suggestions (Optional)

  **1. Error Messaging:**
  When OAuth login fails, users see a generic error. Consider adding user-friendly error messages:
  ```python
  # In views or middleware
  except SocialAuthException as e:
      messages.error(request, "OAuth login failed. Please try again or contact support.")

  2. Session Timeout:
  Consider documenting the session timeout behavior. Users should know how long they stay logged in.

  3. Testing Documentation:
  It would be helpful to add a section in the README about:
  - How to set up beta OAuth consumer for testing
  - Environment variables needed
  - Common troubleshooting steps


  Great foundational work! 🚀

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

zache-fi committed:
Merge branch 'main' into feature/117-oauth-login

sergyDwhiz committed:
fix: Resolve Vue.js template syntax conflicts after merge

- Fixed Django/Vue.js template syntax conflicts reintroduced by merge

sergyDwhiz commented:
Hi @xenacode-art, sorry for my late reply. I've been through a lot. 

Thanks so much for the thorough review and testing. I appreciate you taking the time to verify everything works! Your suggestions about error messaging, session timeout documentation and testing are true. I've been working on a fix for the template syntax conflicts that came back after the merge. Pushing it soon. 
Thanks.

sergyDwhiz committed:
style: Remove unused timezone import

sergyDwhiz committed:
Merge remote-tracking branch 'upstream/main' into feature/117-oauth-login

sergyDwhiz committed:
docs: Update CODEBASE_DOCUMENTATION for review_statistics app refactoring

sergyDwhiz committed:
Merge into main

xenacode-art commented:
@sergyDwhiz  glad to have you back  hope you're doing okay.

sergyDwhiz commented:
Thanks bro, and I'm good. Hope same?

sergyDwhiz commented:
We need to do everything to close this PR sooner bro. It's a crucial feature.
